File: legacy/src/game/pinball-knight/skill-runtime.ts
/**
 * SKILL RUNTIME — the live glue between the pure tree (skills.ts) and the
 * game: a memoized aggregate every hot path reads, XP awards, point spending,
 * and the derived max-HP/max-mana the HUD and potions use.
 *
 * The aggregate is recomputed only when a rank is spent or a run resets —
 * playerDamage and the movement code read it every hit/frame, and rebuilding
 * a Record fold there would be waste.
 */
import { state } from "./state";
import { PLAYER_MAX_HP, MANA_MAX, MANA_POOL_FLOOR, ABILITY_RANK_MAX } from "./constants";
import { SKILLS, aggregateSkills, canLearn, grantXp, xpForFloorClear, XP_KILL, XP_KILL_BOSS, type SkillAggregate, type SkillId } from "./skills";
import { legacyBaseModifiers } from "./legacy";
import { abilityRank, abilityRankCost, type AbilityId } from "./abilities";

let cached: SkillAggregate | null = null;

/** The combined skill + legacy effects. Cheap: memoized until invalidated. */
export function skillAgg(): SkillAggregate {
  if (!cached) cached = aggregateSkills(state.skillRanks, legacyBaseModifiers());
  return cached;
}

/** Call after anything that changes ranks or perks (spend, reset, purchase). */
export function invalidateSkillAgg(): void {
  cached = null;
}

/** Max hearts with Iron Heart ranks + legacy scars + any Elixir of Life bump
 * brewed this run (state.bonusMaxHp; see items.ts / applyPotion). */
export function playerMaxHp(): number {
  return PLAYER_MAX_HP + skillAgg().maxHpFlat + state.bonusMaxHp;
}

/**
 * Max mana with Mana Well ranks — and with the Blood Price keystone's −30,
 * which is the first NEGATIVE contributor this pool has ever had.
 *
 * Floored above the most expensive ability in the table. A pool that could not
 * afford any cast would not be a hard build, it would be a build in which the
 * mana bar is decoration and Blood Price is not a choice — and a keystone whose
 * drawback deletes the system it modifies is a bug wearing a design hat.
 */
export function playerManaMax(): number {
  return Math.max(MANA_POOL_FLOOR, MANA_MAX + skillAgg().manaMaxFlat);
}

/**
 * Abilities usable in the Q/E slots: the two defaults plus the skill tree's
 * unlocks. THE ONE FUNNEL for "what can I cast".
 *
 * Cards deliberately do NOT feed this. The skill tree upgrades the PLAYER;
 * cards upgrade the GEAR. An earlier cut let a socketed card grant an ability,
 * which blurred the only line that makes two progression systems worth having.
 */
export function unlockedAbilities(): AbilityId[] {
  const out = [...state.unlockedAbilities];
  for (const a of skillAgg().unlocked) if (!out.includes(a)) out.push(a);
  return out;
}

/**
 * Drop any Q/E binding whose ability is no longer available, and clear its live
 * cooldown.
 *
 * Called after ANYTHING that can change the held weapon — swap, pickup, a weapon
 * breaking at 0 durability. Without it a card-granted ability stays bound to Q
 * after the weapon carrying it leaves your hand, and casting it would be free
 * power from a weapon you are not holding.
 *
 * Returns true if a slot actually changed, so the caller can flag the HUD dirty.
 */
export function syncAbilitySlots(): boolean {
  const ok = unlockedAbilities();
  let changed = false;
  for (const slot of [0, 1] as const) {
    const id = state.abilitySlots[slot];
    if (id && !ok.includes(id)) {
      state.abilitySlots[slot] = null;
      delete state.abilityCd[id];
      changed = true;
    }
  }
  return changed;
}

/** Level-up fanfare is core's business (toast + sting); wired at launch. */
let onLevelUp: ((level: number, points: number) => void) | null = null;
export function setLevelUpHandler(fn: ((level: number, points: number) => void) | null): void {
  onLevelUp = fn;
}

function award(amount: number): void {
  const scaled = amount * skillAgg().xpMult;
  const next = grantXp({ xp: state.charXp, level: state.charLevel, points: state.skillPoints }, scaled);
  state.charXp = next.xp;
  state.charLevel = next.level;
  state.skillPoints = next.points;
  if (next.levelsGained > 0) {
    const p = state.player;
    if (p) state.vfx?.sparks(p.x, 1.2, p.z, 0, 0, 14);
    onLevelUp?.(next.level, next.points);
  }
  state.hudDirty = true;
}

/** XP for a kill — called from combat's killZombie, the single kill funnel. */
export function awardKillXp(boss: boolean): void {
  award(boss ? XP_KILL_BOSS : XP_KILL);
}

/** Debug-panel XP grant — same pipeline, no pretence of a source. */
export function awardDebugXp(amount: number): void {
  award(amount);
}

/** XP for clearing a floor, scaled by its grade — called from descend(). */
export function awardFloorXp(floor: number, grade: string): void {
  award(xpForFloorClear(floor, grade));
}

/**
 * Spend points into an ABILITY instead of a tree node — the same wallet, so
 * every rank bought here is a tree rank not bought.
 *
 * Gated on the ability being UNLOCKED, deliberately: ranks are an investment in
 * something you can actually cast, and letting a player sink six points into a
 * spell they have no node for would be a trap with no feedback. The escalating
 * price (1, 2, 3) is what stops "dump everything into Arcane Pulse" from being
 * strictly correct.
 */
export function spendAbilityRank(id: AbilityId): { ok: boolean; why?: string } {
  if (!unlockedAbilities().includes(id)) return { ok: false, why: "ability not unlocked" };
  const rank = abilityRank(id);
  if (rank >= ABILITY_RANK_MAX) return { ok: false, why: "maxed" };
  const cost = abilityRankCost(rank);
  if (state.skillPoints < cost) return { ok: false, why: `needs ${cost} point${cost === 1 ? "" : "s"}` };
  state.abilityRanks[id] = rank + 1;
  state.skillPoints -= cost;
  state.hudDirty = true;
  return { ok: true };
}

/** Spend one point into a node. Returns ok/why for the menu to flash. */
export function spendSkillPoint(id: SkillId): { ok: boolean; why?: string } {
  const check = canLearn(id, state.skillRanks, state.skillPoints);
  if (!check.ok) return check;
  const def = SKILLS[id];
  state.skillRanks[id] = (state.skillRanks[id] ?? 0) + 1;
  state.skillPoints -= def.cost;
  invalidateSkillAgg();
  // An unlock is only useful if the player can equip it — surface immediately.
  const unlock = def.modifier.unlockAbility;
  if (unlock && !state.unlockedAbilities.includes(unlock)) state.unlockedAbilities.push(unlock);
  state.hudDirty = true;
  return { ok: true };
}

File: Rust/crates/pk-core/src/player/skill_runtime.rs
//! SKILL RUNTIME — the live glue between the pure tree (skills.ts) and the game.
//!
//! Port of `legacy/src/game/pinball-knight/skill-runtime.ts` (158 lines).
//!
//! PORTS: `skill-runtime.ts`

use std::collections::HashMap;

use crate::abilities::{ability_rank, ability_rank_cost, AbilityId};
use crate::constants::skills::{ABILITY_RANK_MAX, MANA_MAX, MANA_POOL_FLOOR};
use crate::skills::{
    aggregate_skills, can_learn, get_skill, grant_xp, xp_for_floor_clear, SkillAggregate,
    SkillModifier, XpState, XP_KILL, XP_KILL_BOSS,
};

pub const PLAYER_MAX_HP: i32 = 6;

#[derive(Clone, Debug, PartialEq)]
pub struct SkillRuntimeState {
    pub skill_ranks: HashMap<String, usize>,
    pub ability_ranks: HashMap<AbilityId, usize>,
    pub bonus_max_hp: i32,
    pub xp: f64,
    pub level: usize,
    pub skill_points: usize,
    pub unlocked_abilities: Vec<AbilityId>,
    pub ability_slots: [Option<AbilityId>; 2],
    pub ability_cd: HashMap<AbilityId, f64>,
    cached_agg: Option<SkillAggregate>,
}

impl Default for SkillRuntimeState {
    fn default() -> Self {
        Self {
            skill_ranks: HashMap::new(),
            ability_ranks: HashMap::new(),
            bonus_max_hp: 0,
            xp: 0.0,
            level: 1,
            skill_points: 0,
            unlocked_abilities: vec![AbilityId::Flippercharge, AbilityId::Arcanepulse],
            ability_slots: [Some(AbilityId::Flippercharge), Some(AbilityId::Arcanepulse)],
            ability_cd: HashMap::new(),
            cached_agg: None,
        }
    }
}

impl SkillRuntimeState {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn invalidate_skill_agg(&mut self) {
        self.cached_agg = None;
    }

    pub fn skill_agg(&mut self, base: Option<&SkillModifier>) -> SkillAggregate {
        if let Some(ref agg) = self.cached_agg {
            return agg.clone();
        }
        let agg = aggregate_skills(&self.skill_ranks, base);
        self.cached_agg = Some(agg.clone());
        agg
    }

    pub fn player_max_hp(&mut self, base: Option<&SkillModifier>) -> i32 {
        let agg = self.skill_agg(base);
        PLAYER_MAX_HP + agg.max_hp_flat + self.bonus_max_hp
    }

    pub fn player_mana_max(&mut self, base: Option<&SkillModifier>) -> i32 {
        let agg = self.skill_agg(base);
        (MANA_MAX + agg.mana_max_flat).max(MANA_POOL_FLOOR)
    }

    pub fn unlocked_abilities(&mut self, base: Option<&SkillModifier>) -> Vec<AbilityId> {
        let agg = self.skill_agg(base);
        let mut out = self.unlocked_abilities.clone();
        for unl in &agg.unlocked {
            if let Some(id) = AbilityId::from_str_id(unl) {
                if !out.contains(&id) {
                    out.push(id);
                }
            }
        }
        out
    }

    pub fn sync_ability_slots(&mut self, base: Option<&SkillModifier>) -> bool {
        let ok = self.unlocked_abilities(base);
        let mut changed = false;
        for slot in 0..2 {
            if let Some(id) = self.ability_slots[slot] {
                if !ok.contains(&id) {
                    self.ability_slots[slot] = None;
                    self.ability_cd.remove(&id);
                    changed = true;
                }
            }
        }
        changed
    }

    pub fn award(&mut self, amount: f64, base: Option<&SkillModifier>) -> usize {
        let xp_mult = self.skill_agg(base).xp_mult;
        let res = grant_xp(
            &XpState {
                xp: self.xp,
                level: self.level,
                points: self.skill_points,
            },
            amount * xp_mult,
        );
        self.xp = res.xp;
        self.level = res.level;
        self.skill_points = res.points;
        res.levels_gained
    }

    pub fn award_kill_xp(&mut self, boss: bool, base: Option<&SkillModifier>) -> usize {
        self.award(if boss { XP_KILL_BOSS } else { XP_KILL }, base)
    }

    pub fn award_floor_xp(&mut self, floor: u32, grade: &str, base: Option<&SkillModifier>) -> usize {
        self.award(xp_for_floor_clear(floor, grade), base)
    }

    pub fn spend_ability_rank(
        &mut self,
        id: AbilityId,
        base: Option<&SkillModifier>,
    ) -> Result<(), &'static str> {
        let unlocked = self.unlocked_abilities(base);
        if !unlocked.contains(&id) {
            return Err("ability not unlocked");
        }
        let rank = ability_rank(id, &self.ability_ranks);
        if rank >= ABILITY_RANK_MAX {
            return Err("maxed");
        }
        let cost = ability_rank_cost(rank);
        if self.skill_points < cost {
            return Err("not enough skill points");
        }
        self.ability_ranks.insert(id, rank + 1);
        self.skill_points -= cost;
        Ok(())
    }

    pub fn spend_skill_point(&mut self, id: &str) -> Result<(), &'static str> {
        if !can_learn(id, &self.skill_ranks, self.skill_points) {
            return Err("prerequisites or points not met");
        }
        let def = match get_skill(id) {
            Some(d) => d,
            None => return Err("unknown skill node"),
        };
        let current_rank = *self.skill_ranks.get(id).unwrap_or(&0);
        self.skill_ranks.insert(id.to_string(), current_rank + 1);
        self.skill_points -= def.cost;
        self.invalidate_skill_agg();

        if let Some(unl) = def.modifier.unlock_ability {
            if let Some(aid) = AbilityId::from_str_id(unl) {
                if !self.unlocked_abilities.contains(&aid) {
                    self.unlocked_abilities.push(aid);
                }
            }
        }

        Ok(())
    }
}
